File: 0102-binary-tree-level-order-traversal/0102-binary-tree-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        if not root: return []
        queue = [root]
        ans = []
        
        while queue:
            curLevel = []
            
            for i in range(len(queue)):
                curNode = queue.pop(0)
                
                if curNode.left:
                    queue.append(curNode.left)
                
                if curNode.right:
                    queue.append(curNode.right)
                    
                curLevel.append(curNode.val)
                
            ans.append(curLevel)
            
        return ans
        
        

File: 0102-binary-tree-level-order-traversal/0102-binary-tree-level-order-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrder = function(root) {
    //bfs
    /*
    1. check edge case
    2. init a queue and ans array
    3. while queue exists
        3.1) init a curLevel array 
        3.2) init a cur queue length 
        3.3) make a loop to traverse each side. in a loop, get a cur node from queue and put each side into curLevel arr if exist
        3.4) put tue curLevel arr into ans arr
    
    */
    
    if(!root) return [];
    let queue = [root];
    let ans = [];
    
    while(queue.length){
        let curLevel = [];
        let queLevel = queue.length;
        
        for(let i=0; i<queLevel; i++){
            let curNode = queue.shift();
            if(curNode.left) queue.push(curNode.left);
            if(curNode.right) queue.push(curNode.right);
            
            curLevel.push(curNode.val)
        }
        ans.push(curLevel)
    }
    
    return ans
};